tests/upload_test.php

```diff
@@ -957,4 +958,209 @@ public function test_upload_attendance_for_all_session_statuses(int $timestart,
+        ];
+        $session = $generator->create_session($record);
+
+        // Replace contents placeholders.
+        $contents = str_replace(':email', $student->email, $contents);
+        $contents = str_replace(':session', $session->id, $contents);
+
+        // Write csv to stored_file
+        // booking_manager expects it to be in the users draft files.
+        $fs = new file_storage();
+        $storedfile = $fs->create_file_from_string([
+            'contextid' => \context_user::instance($USER->id)->id,
+            'component' => 'user',
+            'filearea' => 'draft',
+            'itemid' => 0,
+            'filepath' => '/',
+            'filename' => uniqid() . '.csv',
+        ], $contents);
+        $bm = new booking_manager($facetoface->id);
+        $bm->load_from_file($storedfile->get_itemid());
+        return [$bm, $session, $student];
+    }
+
+    /**
+     * Test booking_manager validate function
+     *
+     * @param string $contents csv contents
+     * @param string|null $expectedexceptionmessage if given, expect moodle_exception with this message.
+     *
+     * @dataProvider csv_validation_provider
+     */
+    public function test_csv_validation(string $contents, ?string $expectedexceptionmessage = null): void {
+        [$bm, $session, $user] = $this->setup_csv_test($contents);
+        if (!empty($expectedexceptionmessage)) {
+            $this->expectException(\moodle_exception::class);
+            $this->expectExceptionMessage($expectedexceptionmessage);
+        }
+        $bm->validate();
+    }
+
+    /**
+     * Provides CSV processing test cases
+     * @return array
+     */
+    public static function csv_process_provider(): array {
+        // Build a matrix of status, date (past present future and unknown).
+        $dateoptions = [
+            'past' => ['timestart' => time() - DAYSECS * 2, 'timefinish' => time() - DAYSECS],
+            'present' => ['timestart' => time() - DAYSECS, 'timefinish' => time() + DAYSECS],
+            'future' => ['timestart' => time() + DAYSECS, 'timefinish' => time() - DAYSECS * 2],
+            'unknown' => [],
+        ];
+
+        $statusoptions = booking_manager::get_allowed_session_statuses();
+        $tests = [];
+
+        foreach ($dateoptions as $datelabel => $dates) {
+            foreach ($statusoptions as $status) {
+                // Cancelled status only applies to future dates.
+                if ($status == 'cancelled' && $datelabel != 'future') {
+                    continue;
+                }
+
+                // If no status is given, it should be either waitlisted or booked
+                // depending on the time of the session.
+                $expectedstatus = $status;
+                if ($status == '' && $datelabel == 'unknown') {
+                    $expectedstatus = 'waitlisted';
+                } else if ($status == '' && $datelabel !== 'unknown') {
+                    $expectedstatus = 'booked';
+                }
```

Comment:
I think generally speaking in php-land, it's preferable to use `===` instead of `==` where possible. The exception would be if type coercion is desired. 

Alternatively using a method like `is_null` or in this case, `empty(...)` would work too.